akka-jackson/src/test/scala/akka/jackson/JacksonSerializerSpec.scala

```diff
@@ -0,0 +1,204 @@
+import java.time.temporal.ChronoUnit
+import java.util.Arrays
+
+import akka.actor.ActorSystem
+import akka.serialization.SerializationExtension
+import akka.testkit.TestKit
+import com.typesafe.config.ConfigFactory
+import org.scalatest.BeforeAndAfterAll
+import org.scalatest.Matchers
+import org.scalatest.WordSpecLike
+import java.util.Optional
+
+import scala.concurrent.duration.FiniteDuration
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.IntNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+
+object ScalaTestMessages {
+  final case class SimpleCommand(name: String)
+  final case class SimpleCommand2(name: String, name2: String)
+  final case class OptionCommand(maybe: Option[String])
+  final case class BooleanCommand(published: Boolean)
+  final case class TimeCommand(timestamp: LocalDateTime, duration: FiniteDuration)
+  final case class CollectionsCommand(strings: List[String], objects: Vector[SimpleCommand])
+
+  final case class Event1(field1: String)
+  final case class Event2(field1V2: String, field2: Int)
+}
+
+class ScalaTestEventMigration extends JacksonMigration {
+  override def currentVersion = 3
+
+  override def transformClassName(fromVersion: Int, className: String): String =
+    classOf[ScalaTestMessages.Event2].getName
+
+  override def transform(fromVersion: Int, json: JsonNode): JsonNode = {
+    val root = json.asInstanceOf[ObjectNode]
+    root.set("field1V2", root.get("field1"))
+    root.remove("field1")
+    root.set("field2", IntNode.valueOf(17))
+    root
+  }
+}
+
+class JacksonJsonSerializerSpec extends JacksonSerializerSpec("jackson-json")
+class JacksonCborSerializerSpec extends JacksonSerializerSpec("jackson-cbor")
+class JacksonSmileSerializerSpec extends JacksonSerializerSpec("jackson-smile")
+
+abstract class JacksonSerializerSpec(serializerName: String) extends TestKit(ActorSystem(
+  "JacksonJsonSerializerSpec",
+  ConfigFactory.parseString(s"""
+    akka.jackson.migrations {
+      "akka.jackson.JavaTestMessages$$Event1" = "akka.jackson.JavaTestEventMigration"
+      "akka.jackson.JavaTestMessages$$Event2" = "akka.jackson.JavaTestEventMigration"
+      "akka.jackson.ScalaTestMessages$$Event1" = "akka.jackson.ScalaTestEventMigration"
+      "akka.jackson.ScalaTestMessages$$Event2" = "akka.jackson.ScalaTestEventMigration"
+    }
+    akka.actor {
+      allow-java-serialization = off
+      serialization-bindings {
+        "java.lang.Object" = $serializerName
+        "java.io.Serializable" = $serializerName
+      }
+    }
+    """))) with WordSpecLike with Matchers with BeforeAndAfterAll {
+
+  val serialization = SerializationExtension(system)
+
+  override def afterAll {
+    shutdown()
+  }
+
+  def checkSerialization(obj: AnyRef): Unit = {
+    val serializer = serializerFor(obj)
+    val blob = serializer.toBinary(obj)
+    val deserialized = serializer.fromBinary(blob, serializer.manifest(obj))
+    deserialized should ===(obj)
```

Comment:
In addition to checking round-trips, it might be a good idea to test the serialized form against a saved file (sometimes known as [gold master testing](https://codeclimate.com/blog/gold-master-testing/)) in the project, to ensure there are no unexpected changes in the format over time. (Or you could hash it and check against a constant, but at least for text formats it's also useful to see any intentional changes in the diff of the gold file.)

Yet another thing we have wanted to do in Lagom but haven't gotten around to yet...

Reply:
👍

Reply:
@TimMoore FYI I started on something like that in #24631